packages/affiliates/liquidity-mining/CalculateUniswapLPRewards.js

```diff
@@ -0,0 +1,356 @@
+// chunks of `blockPerSnapshot` and at each chunk assign `umaPerSnapshot` at a prorata basis.
+async function _calculatePayoutsBetweenBlocks(
+  uniswapPool,
+  empContract,
+  syntheticToken,
+  shareHolders,
+  fromBlock,
+  toBlock,
+  blocksPerSnapshot,
+  umaPerSnapshot,
+  snapshotsToTake
+) {
+  // Create a structure to store the payouts for all historic shareholders.
+  let shareHolderPayout = {};
+  for (let shareHolder of shareHolders) {
+    shareHolderPayout[shareHolder] = toBN("0");
+  }
+
+  console.log("🏃‍♂️Iterating over block range and calculating payouts...");
+
+  // create new progress bar to show the status of blocks traversed.
+  const progressBar = new cliProgress.SingleBar(
+    {
+      format: "[{bar}] {percentage}% | snapshots traversed: {value}/{total}"
+    },
+    cliProgress.Presets.shades_classic
+  );
+  progressBar.start(snapshotsToTake, 0);
+  for (let currentBlock = fromBlock; currentBlock < toBlock; currentBlock += blocksPerSnapshot) {
+    shareHolderPayout = await _updatePayoutAtBlock(
+      uniswapPool,
+      empContract,
+      syntheticToken,
+      currentBlock,
+      shareHolderPayout,
+      umaPerSnapshot
+    );
+    progressBar.update(Math.ceil((currentBlock - fromBlock) / blocksPerSnapshot) + 1);
+  }
+  progressBar.stop();
+
+  return shareHolderPayout;
+}
+
+// For a given `blockNumber` (snapshot in time), return an updated `shareHolderPayout` object that has appended
+// payouts for a given `uniswapPool` at a rate of `umaPerSnapshot`.
+async function _updatePayoutAtBlock(
+  uniswapPool,
+  empContract,
+  syntheticToken,
+  blockNumber,
+  shareHolderPayout,
+  umaPerSnapshot
+) {
+  // Get the total supply of Uniswap Pool tokens at the given snapshot's block number.
+  const lpTokenSupplyAtSnapshot = toBN(await uniswapPool.methods.totalSupply().call(undefined, blockNumber));
+
+  // Get the total number of synthetics in the Uniswap pool at the snapshot's block number.
+  const syntheticsInPoolAtSnapshot = toBN(
+    await syntheticToken.methods.balanceOf(uniswapPool._address).call(undefined, blockNumber)
+  );
+
+  // Compute how many synthetics each LP token is redeemable for at the current pool weighting.
+  const lpTokensToSynthetics = syntheticsInPoolAtSnapshot.mul(toBN(toWei("1"))).div(lpTokenSupplyAtSnapshot);
+
+  // Get the given holders balance at the given block. Generate an array of promises to resolve in parallel.
+  const uniswapBalanceResults = await Promise.map(
+    Object.keys(shareHolderPayout),
+    shareHolder => uniswapPool.methods.balanceOf(shareHolder).call(undefined, blockNumber),
+    {
+      concurrency: 50 // Keep infura happy about the number of incoming requests.
+    }
+  );
+
+  // Also, get the position information for all shareholders.
+  const tokenShareHolderPositionResults = await Promise.map(
+    Object.keys(shareHolderPayout),
+    shareHolder => empContract.methods.positions(shareHolder).call(undefined, blockNumber),
```

Comment:
is there any issue with someone from shareholder payout not being in either uniswap pool or emp contract?

Reply:
if they are not a uniswap LP then the Graph QL query will not return their address and they wont get iterated over in this step. We effectively start only by looking at uniswap LPs.